### PR TITLE
Wdfo 81 updating sdk for newest version finx api

### PR DESCRIPTION
**Pull request by dk-WZFinTech**
Required by: https://bitbucket.org/dkonwisorz/ownly-backend-fork/pull-requests/461/wdfo-81-updating-sdk-for-newest-version

### CODE REVIEW

**Review by piotrlukaj — APPROVED** [review]
Huge change. I've spotted some things that can be improved. Mostly the language style nitpicks. Overall LGTM 👍

File: figo/figo.py [hunk 1]
```diff
@@ -293,690 +201,822 @@ class FigoConnection(FigoObject):
...
-            'refresh_token': response['refresh_token'] if 'refresh_token' in response else None,
-            'expires': datetime.now() + timedelta(seconds=response['expires_in']),
-            'scope': response['scope'],
+            "access_token": res_data["access_token"],
+            "refresh_token": res_data.get("refresh_token"),
+            "expires": expire_dt,
+            "scope": res_data["scope"],
         }
 
+    # TODO: Missing unit test but used in ownly-backend
```
[the remark this answers is not in the document]
Comment (escer) [reply figo/figo.py:333]: Please remove.

File: figo/figo.py [hunk 2]
```diff
@@ -1,290 +1,198 @@
...
 
 from dotenv import load_dotenv
-load_dotenv()
-
-from datetime import datetime, timedelta
 from requests import Session
-from six import python_2_unicode_compatible
 
-from figo.models import (
+# TODO: We need verify which of this unused import can be remove from SDK
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/figo.py:12]: Threat this ass TODO task

File: figo/figo.py [hunk 3]
```diff
@@ -293,690 +201,822 @@ class FigoConnection(FigoObject):
...
+                "redirect_uri": self.redirect_uri,
+                "grant_type": "authorization_code",
+            },
+            method="POST",
+        )
 
-        return {'access_token': response['access_token'],
-                'refresh_token': response['refresh_token'] if 'refresh_token' in response else None,
-                'expires': datetime.now() + timedelta(seconds=response['expires_in'])}
+        expire_dt = datetime.now() + timedelta(seconds=res_data["expires_in"])
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/figo.py:292]: Will fix that

File: figo/figo.py [hunk 4]
```diff
@@ -293,690 +201,822 @@ class FigoConnection(FigoObject):
...
+            }
+        )
 
-        if 'error' in response:
-            raise FigoException.from_dict(response)
+        res_data = self._request_with_exception(
+            "/auth/token", data, method="POST"
+        )
 
+        expire_dt = datetime.now() + timedelta(seconds=res_data["expires_in"])
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/figo.py:325]: Will fix that

File: figo/models.py [hunk 5]
```diff
@@ -36,58 +28,124 @@ def dump(self):
...
+
+    Attributes:
+        id: internal figo user id
+        full_name: full name
+        email: email address
+        language: two letter code for preferred language
+        created_at: created datetime
+    """
+
+    # TODO: "email" and "password" can be also modify - should we add them here
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/models.py:43]: Sure, will do that

File: tests/test_integrations.py [hunk 6]
```diff
@@ -1,144 +1,211 @@
...
+
+from .conftest import PASSWORD
 
-from dotenv import load_dotenv
 load_dotenv()
 
-API_ENDPOINT = os.getenv("API_ENDPOINT")
-CLIENT_ID = os.getenv("CLIENT_ID")
-CLIENT_SECRET = os.getenv("CLIENT_SECRET")
+# DEMO_PUPPETEER_2_BANK_ID (BLZ: "90090078"):
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply tests/test_integrations.py:14]: I will extend this test to be able to fetch correct ACCESS_METHOD_ID by using DEMO_PUPPETEER_2_BANK_ID

File: tests/test_integrations.py [hunk 7]
```diff
@@ -1,144 +1,211 @@
...
+CREDENTIALS = {"login_id": "user1", "password": "123456"}
+CONSENT = {
+    "recurring": True,
+    "period": 90,
+    "scopes": ["ACCOUNTS", "BALANCES", "TRANSACTIONS"],
+}
+ACCESS_METHOD_ID = "a08d3799-7c08-4ccb-a1b2-e38aba988266"
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply tests/test_integrations.py:21]: https://github.com/W-Z-FinTech-GmbH/python-figo/pull/5#discussion_r604827225

File: figo/figo.py [hunk 8]
```diff
@@ -293,690 +201,822 @@ class FigoConnection(FigoObject):
...
+    # (https://docs.finx.finleap.cloud/stable/#tag/User-Management):
+
     @property
-    def accounts(self):
-        """
-        An array of `Account` objects, one for each account the user has granted the app access.
-        """
-        return self._query_api_object(Account, "/rest/accounts", collection_name="accounts")
+    def user(self):
+        """Get the current figo Account.
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/figo.py:473]: I will re-change it

File: figo/figo.py [hunk 9]
```diff
@@ -293,690 +201,822 @@ class FigoConnection(FigoObject):
...
-        Args:
-            account_or_account_id: account to be removed or its ID
-        """
-        path = "/rest/accounts/{0}".format(getAccountId(account_or_account_id))
-        self._request_with_exception(path, method="DELETE")
+        for k, v in catalog.items():
+            catalog[k] = self._process_catalog_list(v)
+        return catalog
+
+    # TODO: Missing unit test but used in ownly-backend
```
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/figo.py:516]: treat it as temporary TODO - that will be remove before merge

**Comment by dk-WZFinTech** [comment]
This PR will be update later with newest code.

**Comment by dk-WZFinTech** [comment]
CLosed by mistake

File: .travis.yml [hunk 10]
```diff
@@ -1,8 +1,5 @@
 language: python
 python:
-- '2.7'
-- '3.4'
-- '3.5'
 - '3.6'
```
Comment (escer) [inline .travis.yml:3]: How about testing under 3.8 as well?

File: figo/figo.py [hunk 11]
```diff
@@ -1,290 +1,192 @@
...
+            "Content-Type": "application/json",
+            "User-Agent": f"python_figo/{__version__}",
         }
         self.language = language
         self.api_endpoint = api_endpoint
 
-    def _request_api(self, path, data=None, method="GET", raise_exception=False):
+    @staticmethod
+    def _has_error(response):
+        return "error" in response and response["error"]
```
Comment (escer) [inline figo/figo.py:54]:
```suggestion
        return bool(response.get("error"))
```

File: figo/figo.py [hunk 12]
```diff
@@ -293,690 +195,764 @@ class FigoConnection(FigoObject):
...
+        query_params = urlencode({"country": country_code, "q": item_id})
 
-        return self._query_api_object(Sync, "/rest/accesses/{0}/syncs".format(access_id), data=data, method='POST')
+        # now the catalog returns matches for all possible banks
+        response = self._query_api_object(
+            LoginSettings,
+            f"/catalog/banks?{query_params}",
+            collection_name="collection",
+        )
+        if response and len(response) > 0:
```
Comment (escer) [inline figo/figo.py:528]: Second condition appears to be redundant.
```suggestion
        if response:
```

**Review by escer — APPROVED** [review]
No blockers found :+1:

File: figo/models.py [hunk 13]
```diff
@@ -185,9 +369,9 @@ def get_security(self, security_id):
         """
         return self.session.get_security(self.account_id, security_id)
 
-    def __unicode__(self):
-        return u"Account: %s (%s at %s)" % (
-            self.name, self.account_number, self.bank_name
+    def __str__(self):
+        return (
+            f"Account: {self.name} ({self.account_number} at {self.bank_name})"
```
Comment (piotrlukaj) [inline figo/models.py:374]: I'd use a single quotes:
`f'Account: {self.name} ({self.account_number} at {self.bank_name})'`

here and in other places within this file.

File: tests/test_integrations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (piotrlukaj) [reply tests/test_integrations.py:14]: Is it extended by now? :)

**Review by piotrlukaj — APPROVED** [review]
A one nitpick regarding the quotes is added. One conversation resolved and one left :)

I wonder - should we apply the typehints?

File: figo/exceptions.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
+ERROR_MESSAGES = {
+    400: {
+        "message": "bad request",
+        "description": "Bad request",
+        "code": 90000,
```
[the remark this answers is not in the document]
Comment (piotrlukaj) [reply figo/exceptions.py:5]: Did you verified it?

File: .travis.yml (hunk 10 above)
Comment (dk-WZFinTech) [reply .travis.yml:3]: I will move this to new issue: #6

File: figo/models.py (hunk 13 above)
Comment (dk-WZFinTech) [reply figo/models.py:374]: In this project from beginning dominated was double-quotes, so I will leave them as they are.

File: figo/exceptions.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply figo/exceptions.py:5]: I didn't have a time to do it. I will move this problem to new issue: #7

File: tests/test_integrations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dk-WZFinTech) [reply tests/test_integrations.py:14]: Moved to #8

**Comment by dk-WZFinTech** [comment]
> I wonder - should we apply the typehints?

It is good idea. But I will move it also to separate issue: #9

File: figo/figo.py (hunk 11 above)
Comment (dk-WZFinTech) [reply figo/figo.py]: This change broke test:

```python
---------------------------------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------------------------------
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): staging.figo.me:443
DEBUG:urllib3.connectionpool:https://staging.figo.me:443 "GET /rest/accesses/X677030772078018561.1/syncs/aac860f0-93d3-4f85-90b3-2bb0b87711be/challenges HTTP/1.1" 200 189
DEBUG:figo.figo:GET 'https://staging.figo.me/rest/accesses/X677030772078018561.1/syncs/aac860f0-93d3-4f85-90b3-2bb0b87711be/challenges' result with status 200 and text: [{"id": "9fc8be3c-f1c3-4d79-9696-6728e3312187.1", "created_at": "2021-07-19T08:36:51.348191+00:00", "type": "EMBEDDED", "label": "", "format": "TEXT", "data": "Enter valid OTP to log in."}]
------------------------------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------------------------------
DEBUG    urllib3.connectionpool:connectionpool.py:975 Starting new HTTPS connection (1): staging.figo.me:443
DEBUG    urllib3.connectionpool:connectionpool.py:461 https://staging.figo.me:443 "GET /rest/accesses/X677030772078018561.1/syncs/aac860f0-93d3-4f85-90b3-2bb0b87711be/challenges HTTP/1.1" 200 189
DEBUG    figo.figo:figo.py:90 GET 'https://staging.figo.me/rest/accesses/X677030772078018561.1/syncs/aac860f0-93d3-4f85-90b3-2bb0b87711be/challenges' result with status 200 and text: [{"id": "9fc8be3c-f1c3-4d79-9696-6728e3312187.1", "created_at": "2021-07-19T08:36:51.348191+00:00", "type": "EMBEDDED", "label": "", "format": "TEXT", "data": "Enter valid OTP to log in."}]
===================================================================================================================== short test summary info =====================================================================================================================
FAILED tests/test_integrations.py::test_get_accesses - AttributeError: 'list' object has no attribute 'get'
FAILED tests/test_integrations.py::test_get_synchronization_challenges - AttributeError: 'list' object has no attribute 'get'
============================================================================================================= 2 failed, 53 passed in 63.24s (0:01:03) =============================================================================================================
```

I will revert this commit.

Comment (dk-WZFinTech) [reply figo/figo.py]: Instead of revert commit I use condition:

```python
response = response[0] if isinstance(response, list) else response
```

**Pull request merged**